Introduce new CodeMirror block for editing code.

dmsnell opened the pull request:
While demonstrating Gutenberg to my team the question was asked about
editing code blocks: "it's so terribly difficult in the WordPress editor."

Wanting to demonstrate some of the flexibility that the block system
provides I created a new block type which turns a normal block into an
instance of the CodeMirror editor. 

Issues:
  - How do we determine which modes to show? I think we should allow for all of them but have a better way to select the common modes. Maybe an auto-completing drop-down menu in the inspector.
  - Travis fails on CodeMirror's CSS but it runs locally

![codemirrorblock](https://user-images.githubusercontent.com/5431237/31186809-be00181c-a8fd-11e7-96ab-c220a4f8b7d0.gif)


Also shown here with a degraded network connection to show the async code loading in when requested
![codemirrorlazyloading](https://user-images.githubusercontent.com/5431237/31187344-9eeb9044-a8ff-11e7-9be0-0f6d91d05809.gif)

dmsnell commented:
does anyone care about this? any reason to keep it around?

BE-Webdesign commented:
> does anyone care about this? any reason to keep it around?

I think this would be an awesome thing to have in the editor.  This could also potentially become a good test case for building a plugin that registers a block.  I doubt the vast majority of WordPress users will need/want this kind of content block, but any programmer would probably really like this!

dmsnell commented:
thanks @BE-Webdesign - your comment makes me realize that I should create this at some point outside of the project as a new-editor plugin!~

closing for now then

dmsnell commented:
Reopening for growing interest

iseulde commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,62 @@
+const { html } = query;
+
+const toEditor = s => he.decode( s.split( '<br>' ).join( '\n' ) );
+const fromEditor = s => he.encode( s, { useNamedReferences: true } ).split( '\n' ).join( '<br>' );
```
If you use `text` to query for the attribute, and `<pre><code>`, this is all not needed? React will do it by itself? That's what I did with the plain code block.

dmsnell replied:
…will investigate, thanks!

have you not had any problems with special characters or whitespace? can I get rid of `he` altogether?

(note also that `he` has a few cool points, like being able to choose how and when characters get encoded)

iseulde replied:
I think you can remove it entirely, yes. It will convert to named entities.

aduth replied:
>(note also that `he` has a few cool points, like being able to choose how and when characters get encoded)

It's also a behemoth of a library, at least 24kb minified + gzipped. Might be a point toward dynamically enqueueing block scripts separately based on presence in content.

dmsnell replied:
thanks! I pulled it in initially because it was a prototype. I'm very glad you were aware of how `<pre><code>` changed the semantics. much better now!

iseulde commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,62 @@
+	edit( { attributes, setAttributes } ) {
+		return (
+			<div>
+				<select onBlur={ ( { target: { value } } ) => setAttributes( { language: value } ) }>
```
Wouldn't `onChange` be better?

dmsnell replied:
had `onChange` to begin with but `eslint` yapped at me.

dmsnell replied:
used `onChange` - please see my new comment asking about this…

iseulde commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,62 @@
+					{ [ 'css', 'diff', 'javascript', 'markdown', 'pegjs', 'php' ].map( language => (
+						<option
+							key={ language }
+							selected={ language === attributes.language }
```
Getting a warning:

> Warning: Use the `defaultValue` or `value` props on `<select>` instead of setting `selected` on `<option>`.

dmsnell replied:
thanks. I honestly didn't pay any attention to this (obviously). it seems like I need to make it a setting in the page inspector too but at the time I didn't know about that

dmsnell replied:
used `value`

iseulde commented on post-content.js:
```diff
@@ -163,6 +163,10 @@ window._wpGutenbergPost = {
 			'https://make.wordpress.org/core/2017/01/17/editor-technical-overview/',
 			'<!-- /wp:core/embed -->',
 
+			'<!-- wp:core/code-mirror language="php" -->',
```
JS? :)

dmsnell replied:
…will change to Elm :wink:

dmsnell replied:
changed to Elm :smile:

dmsnell commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,65 @@
+						<label className="blocks-text-control__label" htmlFor="blocks-codemirror-language-select">{ __( 'Language' ) }</label>
+						<select // eslint-disable-line jsx-a11y/no-onchange
+							id="blocks-codemirror-language-select"
+							onChange={ ( { target: { value } } ) => setAttributes( { language: value } ) }
```
@iseulde I changed this back to `onChange` because `onBlur` wasn't updating the setting until I clicked outside of the box. now I'm in a quandary: `eslint` wants `onBlur` but it appears like we have to use `onChange` to get the highlighting to update with the selection.

any thoughts?

dmsnell commented:
I've got some difficulty understanding the CSS errors. They are coming in via CodeMirror.

Honestly I would prefer to be able to dynamically load the CSS but I'm not sure how that would happen. if someone would be willing to help me through these errors/async issues I'll buy you an 🍦

dmsnell commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,65 @@
+require( 'codemirror/mode/markdown/markdown' );
+require( 'codemirror/mode/pegjs/pegjs' );
+require( 'codemirror/mode/php/php' );
+require( 'codemirror/lib/codemirror.css' );
```
there are more of these. if we had a way to async load them I would do so on-demand and provide the full suite with a better language selector

gziolo replied:
Did you try using `import` instead of `require` here? It should teach Webpack to async load all those files. I guess we would still have to wrap `CodeMirror` into its own component to make sure those deps are loaded only when it mounts.

dmsnell replied:
I tried using `import()` and it didn't seem to recognize it. In d957054 I added the use of `require.ensure()` which seems to mostly work. There is still an issue that it's bundling the entirety of the modes directory into one chunk instead of many.

westonruter commented:
FYI: @georgestephanis has a feature plugin for introducing CodeMirror into other areas of WP. The block perhaps could be made part of that plugin and re-use the copy of CodeMirror that it includes.

See https://github.com/georgestephanis/codemirror-wp

georgestephanis commented:
Just for a separation of concerns, and to keep Gutenberg a little bit lighter, it might be wise for the Gutenberg version of a code editor block to just be a plain textarea, and have the CodeMirror-WP plugin handle turning it more interactive and engaging?

I'm happy to add anyone to the cm repo, just ask.

dmsnell commented:
> and have the CodeMirror-WP plugin handle turning it more interactive and engaging?

thanks @georgestephanis. the block editor is a very good place, in my opinion, to pull in code mirror. since it _is_ implementation specific I would imagine a WordPress plugin that both registers the editor block _and_ includes the script into the browser.

this PR isn't going to be considered ready regardless until we can asynchronously load in CodeMirror and its modes. it would keep Gutenberg lightweight and make code editing first-class.

actually I would prefer to generalize it as something like "syntax editor" or "code editor" in its name so we could swap out CodeMirror if we wanted.

georgestephanis commented:
Sure. Naming is just what I'd thrown on at the time. If you'd like to have the block registered as part of the cm plugin, that's fine too. It's all aimed at going the same place in the end. 👍

mtias commented:
See also https://github.com/ataylorme/gutenberg/pull/1

dmsnell commented:
Any suggestions on where to move this branch? I've closed it once and I don't mind closing it again, but I hate to see it just sour here without movement.

mtias commented:
Do we have async loading of the library yet?

dmsnell commented:
> Do we have async loading of the library yet?

ah no! good point - this was waiting on me 🙈

gziolo reviewed (commented):
> Do we have async loading of the library yet?

Is this the only remaining blocker here? Do we want to have it in core or rather as a plugin? I'm happy to help with this one.

gziolo commented:
> Reopening for growing interest

I'm not surprised :) Using CodeMirror was my first thought when I saw the existing code block 😄

iseulde commented:
If core adds it, let's add it to the default code block, but also keep a "plain" type (no language). Also note we'll need highlighting on the front-end. [Prism](http://prismjs.com) is quite nice.

mtias commented:
We used it for the docs: http://gutenberg-devdoc.surge.sh/block-api/

With a tweaked stylesheet.

mtias commented:
And agreed that if it lands in core we should make it part of the main block.

dmsnell commented:
This has been updated to lazy-load CodeMirror and also the language modes. There are some problems indicated inside the code itself on those. Notably:
  - When dynamically loading the CSS it ends up rendering wrong so I have left the CSS as a static import
  - The mode files are all being bundled into one file instead of having separate chunks for each mode. This is related to `require.context()` and webpack statically anticipating the import. I believe that by doing some twiddling we can get what we want. I'm having a hard time really understanding what webpack is doing.

At first I tried manually inserting a new `<script>` tag into the document head, which was loading files as requested, but the paths were screwy and I had issues with the loaded files expecting `CodeMirror` to be a global in a specific state which didn't make sense. Ideally I think we'd rely on webpack to handle the glue for us.

The tests are still failing and I didn't really bother to look at why. The PR has known issues to address before fixing the tests.

gziolo commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,128 @@
+// @see node_modules/codemirror/modes
+// when supported we can probably replace with standard `import()` syntax
+const modes = {
+	apl: { label: 'APL', req: () => System.import( 'codemirror/mode/apl/apl' ) },
```
`System.import` is deprecated. You should use dynamic `import` instead`. See webpack 3 announcement post: https://medium.com/webpack/webpack-3-official-release-15fd2dd8f07b:

```js
import(/* webpackChunkName: "my-chunk-name" */ 'module');
```

gziolo replied:
I just noticed we are on Webpack 2.7, but I think it still supports `import`, but without this magic comments.

dmsnell replied:
See the comment. `import()` was throwing errors as not supported

dmsnell replied:
I can double check though too

gziolo replied:
Maybe we miss something in Babel config  ¯\_(ツ)_/¯

dmsnell replied:
@gziolo and @aduth I'm getting this when trying to use `import()`

```
ERROR in ./blocks/library/code-mirror/index.js
Module build failed: SyntaxError: Unexpected token (24:33)
```

Not seeing a source of that error either.

dmsnell replied:
https://github.com/babel/babel-loader/issues/493

gziolo replied:
https://github.com/tc39/proposal-dynamic-import - it's stage 3. It learned yesterday that Gutenberg uses only features that are accepted by ECMA TC39. So this one needs to wait. It's good as it is :)

dmsnell committed:
Introduce new CodeMirror block for editing code.

While demonstrating Gutenberg to my team the question was asked about
editing code blocks: "it's so terribly difficult in the WordPress
editor."

Wanting to demonstrate some of the flexibility that the block system
provides I created a new block type which turns a normal block into an
instance of the CodeMirror editor.

This should eventually replace the standard `code` block once it's ready
for production. At the moment it lazy-loads the `CodeMirror` editor
_and_ the mode files, so thankfully this keeps the initial builds small.

gziolo commented on blocks/library/code-mirror/index.js:
```diff
@@ -0,0 +1,128 @@
+// when this CSS was imported dynamically
+// it was causing major visual glitches to
+// the editor which completely broke it
+require( 'codemirror/lib/codemirror.css' );
```
All styles are loaded using the following syntax:

```js
import './style.scss';
```

Webpack uses the following pattern `test: /style\.s?css$/,` to match styles. So it should work here, too. Unless this is the reason why it breaks :)

dmsnell replied:
actually I think it had to do with race conditions on the mounted components. I was comparing to async loading as the dynamic bits though vs. `import` statement.

mainly I meant to convey that this stylesheet seemed to need to load before the editor and I wasn't sure how best to do that

gziolo commented:
> The tests are still failing and I didn't really bother to look at why. The PR has known issues to address before fixing the tests.

Yes, it's enforced to add fixtures for all blocks in here: https://github.com/WordPress/gutenberg/blob/master/blocks/test/full-content.js#L198.

dmsnell committed:
Update with project

karmatosed commented:
As talked about in this week's chat, closing this as now Codemirror is in for one block it needs iterating.

gziolo commented:
https://github.com/WordPress/gutenberg/blob/master/components/code-editor/index.js - we have CodeMirror integrated here, we should try to use a similar approach or even better extend the existing one to make it more flexible.

dmsnell commented:
😢 

this was my favorite block
